src/connection/socket-connection.ts

```diff
@@ -0,0 +1,215 @@
+    clearInterval(this.heartbeatInterval);
+    // clean up all socket event listeners
+    this.socket.removeAllListeners();
+
+    const closePromises = [];
+    for (const [id, subscription] of this.subscriptions) {
+      closePromises.push(subscription.close());
+      this.subscriptions.delete(id);
+    }
+
+    // close all of the associated subscriptions
+    await Promise.all(closePromises);
+
+    // close the socket.
+    this.socket.close();
+  }
+
+  /**
+   * Pong messages are automatically sent in response to ping messages as required by
+   * the websocket spec. So, no need to send explicit pongs.
+   */
+  private pong(): void {
+    this.isAlive = true;
+  }
+
+  /**
+   * Log the error and close the connection.
+   */
+  private async error(error:Error): Promise<void>{
+    log.error(`SocketConnection error, terminating connection`, error);
+    this.socket.terminate();
+    await this.close();
+  }
+
+  /**
+   * Handles a `JSON RPC 2.0` encoded message.
+   */
+  private async message(dataBuffer: Buffer): Promise<void> {
+    const requestData = dataBuffer.toString();
+    if (!requestData) {
+      return this.send(createJsonRpcErrorResponse(
+        uuidv4(),
+        JsonRpcErrorCodes.BadRequest,
+        'request payload required.'
+      ))
+    }
+
+    let jsonRequest: JsonRpcRequest;
+    try {
+      jsonRequest = JSON.parse(requestData);
+    } catch(error) {
+      const errorResponse = createJsonRpcErrorResponse(
+        uuidv4(),
+        JsonRpcErrorCodes.BadRequest,
+        (error as Error).message
+      );
+
+      return this.send(errorResponse);
+    };
+
+    const requestContext = await this.buildRequestContext(jsonRequest);
+    const { jsonRpcResponse } = await jsonRpcApi.handle(jsonRequest, requestContext);
+    if (jsonRpcResponse.error) {
+      requestCounter.inc({ method: jsonRequest.method, error: 1 });
+    } else {
+      requestCounter.inc({
+        method: jsonRequest.method,
+        status: jsonRpcResponse?.result?.reply?.status?.code || 0,
+      });
+    }
+    this.send(jsonRpcResponse);
+  }
+
+  /**
+   * Sends a JSON encoded Buffer through the Websocket. Accepts an error handler, if none is provided the error is logged.
+   */
+  private send(response: JsonRpcResponse | JsonRpcErrorResponse, onError?: (error?: Error) => void): void {
+    if (!onError) {
+      onError = (error):void => {
+        if(error) { log.error('socket send error', error, response); }
+      }
+    }
+
+    this.socket.send(Buffer.from(JSON.stringify(response)), onError);
+  }
+
+  /**
+   * Subscription Handler used to build the context for a `JSON RPC` API call.
+   * Wraps the incoming `message` in a `JSON RPC Success Response` using the original subscription`JSON RPC Id` to send through the WebSocket.
+   */
+  private subscriptionHandler(id: JsonRpcId): (message: GenericMessage) => void {
+    return (message) => {
+      const response = createJsonRpcSuccessResponse(id, { reply: {
```

Comment:
Confirming my understanding: so `id` here is intended to be the mechanism for the client to identify what "subscription" the message is for correct?

And that an alternative for such identification would be to pass `target` and `messageCid` instead into the `subscriptionHandler()` so it can come out in the response right?

Reply:
> And that an alternative for such identification would be to pass target and messageCid instead into the subscriptionHandler() so it can come out in the response as result right?

For the most part, yes. But We'd get rid of JsonRpcClient in that case since it is kind of a 'dumb' client currently that doesn't know about the DWN message structure and just subscribes to a json rpc request and identifies that subscription based on that.